cmd/config/config.go

```diff
@@ -0,0 +1,298 @@
+)
+
+// GetKVS - get kvs from specific subsystem.
+func GetKVS(s string, scfg Config) (map[string]KVS, error) {
+	if len(s) == 0 {
+		return nil, Error("input cannot be empty")
+	}
+	inputs := strings.Fields(s)
+	if len(inputs) > 1 {
+		return nil, Error(fmt.Sprintf("invalid number of arguments %s", s))
+	}
+	subSystemValue := strings.SplitN(inputs[0], subSystemSeparator, 2)
+	var found bool
+	for _, v := range SubSystems {
+		if v == subSystemValue[0] {
+			found = true
+			break
+		}
+		// Check for sub-prefix only if the input value
+		// is only a single value, this rejects invalid
+		// inputs if any.
+		if strings.HasPrefix(v, subSystemValue[0]) && len(subSystemValue) == 1 {
+			found = true
+			break
+		}
+		found = false
+	}
+	if !found {
+		return nil, Error(fmt.Sprintf("unknown sub-system %s", subSystemValue[0]))
+	}
+
+	kvs := make(map[string]KVS)
+	var ok bool
+	if len(subSystemValue) == 2 {
+		kvs[inputs[0]], ok = scfg[subSystemValue[0]][subSystemValue[1]]
+		if !ok {
+			err := fmt.Sprintf("sub-system target %s doesn't exist, feel free to create a new one", s)
+			return nil, Error(err)
+		}
+		return kvs, nil
+	}
+	for subSys, subSysTgts := range scfg {
+		if !strings.HasPrefix(subSys, subSystemValue[0]) {
+			continue
+		}
+		for k, kv := range subSysTgts {
+			if k != Default {
+				kvs[subSys+subSystemSeparator+k] = kv
+			} else {
+				kvs[subSys] = kv
+			}
+		}
+	}
+	return kvs, nil
+}
+
+// DelKVS - delete a specific key.
+func DelKVS(s string, scfg Config) error {
+	if len(s) == 0 {
+		return Error("input arguments cannot be empty")
+	}
+	inputs := strings.Fields(s)
+	if len(inputs) > 1 {
+		return Error(fmt.Sprintf("invalid number of arguments %s", s))
+	}
+	subSystemValue := strings.SplitN(inputs[0], subSystemSeparator, 2)
+	var found bool
+	for _, v := range SubSystems {
+		if v == subSystemValue[0] {
+			found = true
+			break
+		}
+		found = false
+	}
+	if !found {
+		return Error(fmt.Sprintf("unknown sub-system %s", subSystemValue[0]))
+	}
+	if len(subSystemValue) == 2 {
+		delete(scfg[subSystemValue[0]], subSystemValue[1])
+		return nil
+	}
+	return Error("default config cannot be removed")
+}
+
+// SetKVS - set specific key values per sub-system.
+func SetKVS(s string, scfg Config) error {
+	if len(s) == 0 {
+		return Error("input arguments cannot be empty")
+	}
+	inputs := strings.Fields(s)
+	if len(inputs) <= 1 {
+		return Error(fmt.Sprintf("invalid number of arguments %s", s))
+	}
+	subSystemValue := strings.SplitN(inputs[0], subSystemSeparator, 2)
+	var kvs = KVS{}
+	for _, v := range inputs[1:] {
+		kv := strings.SplitN(v, kvSeparator, 2)
+		if len(kv) == 0 {
+			continue
+		}
+		kvs[kv[0]] = kv[1]
+	}
+
+	var found bool
+	for _, v := range SubSystems {
+		if v == subSystemValue[0] {
+			found = true
+			break
+		}
+		found = false
+	}
+	if !found {
+		return Error(fmt.Sprintf("unknown sub-system %s", subSystemValue[0]))
+	}
+	if len(subSystemValue) == 2 {
+		scfg[subSystemValue[0]][subSystemValue[1]] = kvs
```

Comment:
Why do we override the entire sub-system upon addition? We need to just edit/substitute the key:value of that sub-system right?

For example,

Initally the notify_webhook subsystem looked like this
```
"notify_webhook": {
		"1": {
			"endpoint": "http://localhost:8080/",
			"queue_dir": "/tmp/webhook_events",
			"queue_limit": "0",
			"state": "on"
		}
	}
```
And after this `mc admin config set myminio/ notify_webhook:1 queue_dir="/tmp/events"`. I saw

```
"notify_webhook": {
		"1": {
			"queue_dir": "/tmp/events"
		}
	}
```

And the ARN gets removed automatically as the config is affected

Reply:
And i was able to set invalid kv 

```
"notify_webhook": {
		"1": {
			"abc": "def"
		}
	}
```

Reply:
@Praveenrajmani - yes I am adding support for only specific keys alone by looking at the right value. 

Now why a webhook sub-system is allowed because the validation code in webhook doesn't validate the targets correctly, you are allowed to set invalid targets anyways. Enhancing the validation of targets for each sub-system is out of scope for this PR.

This PR should simply move to KV format, no further enhancements to target validations are added. Without scope limits this PR may go on forever.

Reply:
Ok.. So, we need to specify all the other fields as well while setting? like `mc admin config set myminio/ notify_webhook:1 endpoint="http://localhost:8080/" queue_dir="/tmp/events"` queue_limit = "0" state="on"` ?

Reply:
In the change that you have locally yes, I just pushed a new change that adds `comment` field and also allows for toggling individual fields. 

I am continuously pushing changes so things would change rapidly.